DietPi-DDNS | Manage domains for your dynamic IP

nonnorm opened the pull request:
**Status**: Review

**References**:
- https://github.com/MichaIng/DietPi/issues/2868
- https://dietpi.com/phpbb/viewtopic.php?t=8041

**ToDo**:
- [x] Replace inline cURL with dedicated script to avoid having plain text passwords/tokens in journalctl logs
- [x] Replace password input box with password box (which replaces typed characters with `*`)? Enhances security but especially with for those long tokens it's a comfort downside. ... Ah, it's the very same with UNIX passwords, let's be consequent here. Via SSH copy&paste is always possible anyway.

nonnorm committed:
v7.1
+ DietPi-Software | Add DuckDNS

nonnorm committed:
v7.1
+ DietPi-Config | Add DuckDNS to avalible network options

MichaIng commented:
Ay, before you invest too much time into the software install part of this, I'm working on a DietPi-DDNS script with generic DDNS provider support for those which can be updated via simple request, including support to enter a custom provider update URL. Similarly to DietPi-VPN, there is not much point in having this as a software install option as the script will be always present.

I'll incorporate/merge it with our DuckDNS specific work, especially the "OK" return code to check for success is a good idea.

MichaIng commented:
I restored the branch, and re-opened PR to compare/incorporate the work.

nonnorm committed:
v7.1
+ DietPi-Software | Typo

nonnorm committed:
v7.1 (#4262)

+ Update DietPi-DDNS branch

nonnorm committed:
Merge branch 'dev' into duckdns

nonnorm committed:
v7.1
+ DietPi-DDNS | Add DietPi-DDNS script

nonnorm committed:
v7.1
+ DietPi-Config | (Hopefully) CodeFactor Fixes

nonnorm commented:
It's ready for you, @MichaIng.

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -4190,6 +4190,20 @@ Additional benchmarks:
+			duckdns_menutext='[Offline] Invalid DuckDNS details'
+
+			# Check if DuckDNS url returns OK
+			[[ "$(eval $(grep 'duckdns.org/update' /var/lib/dietpi/dietpi-software/installed/duckdns.sh))" == 'OK' ]] && duckdns_menutext='[Online] Change DuckDNS details'
```
@ravenclaw900 
Little not about the CodeFactor error here:
- The left sided double quotes in bash `[[ ]]` is not required, as word splitting doesn't play a role and glob expansion (like allowing `*` and `[0-9]`) is only done on unquoted patterns on the right side of `==`.
- But inside of command substitutions `$(command arguments...)`, word splitting is relevant as with every other command call.
- This is particular case, `eval` will to execute the intended `curl`, as `grep` returns multiple arguments, while eval needs to get everything in a single argument. So the statement needs to be:
    ```
    [[ $(eval "$(grep 'duckdns.org/update' /var/lib/dietpi/dietpi-software/installed/duckdns.sh)") == 'OK' ]]
    ```

But do not bother, I'm adding `dietpi-ddns` after doing another important change to `dietpi-software` from my ToDo.

MichaIng committed:
v7.1

+ DietPi-DDNS | Create new tool to setup background jobs to regularly update dynamic IPs against DDNS providers

MichaIng committed:
v7.1

+ DietPi-Config | Network Options: Misc: Replace No-IP and DuckDNS menu entries with "Dynamic DNS", to start the new dietpi-ddns script

MichaIng committed:
v7.1

+ DietPi-Software | Remove DuckDNS and No-IP install options, now integrated into new dietpi-ddns script

MichaIng committed:
v7.1

+ DietPi-Launcher | Add new DietPi-DDNS

MichaIng committed:
v7.1

+ DietPi-Login | Add "dietpi-ddns" alias

MichaIng committed:
v7.1

+ CHANGELOG | Add notes about DietPi-DDNS and removed DietPi-NordVPN + No-IP install options

MichaIng committed:
v7.1

+ README | Remove No-IP client from 3rd party list

MichaIng committed:
v7.1

+ DietPi-LetsEncrypt | Replace No-IP hint with dietpi-ddns

MichaIng committed:
v7.1

+ DietPi-Survey_report | Remove No-IP support from v7.1 on

MichaIng committed:
Merge branch 'dev' into duckdns

MichaIng committed:
v7.1

+ DietPi-Survey_report | Satisfy shellcheck

MichaIng committed:
v7.1

+ DietPi-DDNS | Remove obsolete No-IP client, when found, before setting up Cron job

MichaIng committed:
v7.1

+ DietPi-DDNS | Use separate update script which holds the cURL call with credentials and has strict read-only permissions. This assures that credentials are never shown as part of Cron journalctl logs.
+ DietPi-DDNS | Use password box for password/token input. While it makes the input harder, it enhances security by assuring that those are never visible on screen.

nonnorm committed:
v7.1
+ DietPi-DDNS | Use `noip.com` instead of `no-ip.com`

nonnorm committed:
Merge branch 'duckdns' of https://github.com/MichaIng/DietPi into duckdns

MichaIng committed:
v7.1

+ DietPi-Patches | Remove obsolete DietPi-NordVPN and No-IP install states

Joulinar commented:
Question: what happen with user who have NoIP installed as of today. Are they able to continue to mange their NoIP account? Or do they need to re-setup thinks using `dietpi-ddns` or are we going to provide a migration path? I mean we are going to remove the install option, means people are not able to uninstall via `dietpi-software` anymore. Probably I missed something as I was checking the code just briefly

MichaIng committed:
v7.1

+ DietPi-Patches | Fix install state removal, especially on first boot where the states file does not exist yet

MichaIng commented:
AFAIK `/usr/local/etc/no-ip.conf` contains the password in a hashed form, so there is no way for us to do an automated migration. The No-IP client stays installed and functional. Only when using `dietpi-ddns` to setup and finally apply a new DDNS Cron job, and the update test succeeds, the No-IP client + service is removed. The setup would need to be done manually indeed.

What we should add is an interactive prompt to inform and encourage users with No-IP installed to setup No-IP DDNS via our new script, which is lighter and uses up-to-date protocols etc and have the old client automatically removed as very last step of the setup. Most importantly it cannot happen that the client is removed but no Cron job installed.

MichaIng committed:
v7.1

+ DietPi-Patches | Add info about No-IP client migration

MichaIng commented:
That should make it clear: https://github.com/MichaIng/DietPi/pull/4229/commits/5c83d4eed7bb5f56c510ad1511345c762acc842e

Joulinar commented:
Yeah that should be fine 😉

MichaIng committed:
v7.1

+ DietPi-Patches | Syntax

MichaIng committed:
v7.1

+ DietPi-DDNS | Do not use sudo to run DDNS update test, to avoid plain text passwords/tokens in journal sudo log. The chance that the unprivileged user causes an issue is pretty low. That would require a strict custom setup/firewall to permit network access only for specific users, which would break various other common implementations as well. So it's okay to count on every user being able to perform network requests.

MichaIng committed:
v7.1

+ DietPi-DDNS | Current settings now need to be read from the new script as well

MichaIng committed:
v7.1

+ DietPi-DDNS | Fix obtaining time span

MichaIng committed:
v7.1

+ DietPi-DDNS | Workaround shellcheck false positive

MichaIng commented on dietpi/dietpi-ddns:
```diff
@@ -80,7 +80,7 @@ Read()
 
 	# No-IP
-	elif [[ $cronjob == *'no-ip.com'* ]]
+	elif [[ $cronjob == *'noip.com'* ]]
```
@ravenclaw900 
The API docs say "no-ip.com": https://www.noip.com/integrate/request
But indeed both work. They are pointing to the same IP, just slightly different TTL:
- https://dnschecker.org/all-dns-records-of-domain.php?query=dynupdate.no-ip.com&rtype=ALL&dns=cloudflare
- https://dnschecker.org/all-dns-records-of-domain.php?query=dynupdate.noip.com&rtype=ALL&dns=cloudflare

nonnorm commented on dietpi/dietpi-ddns:
```diff
@@ -0,0 +1,372 @@
+	if [[ $PROVIDER == 'DuckDNS' ]]
+	then
+		url="https://www.duckdns.org/update?domains=$DOMAINS&token=$PASSWORD"
+		unset -v USERNAME PASSWORD
```
> unset -v USERNAME PASSWORD

Any reason for this line?  Just asking because it makes the 'Token' field disappear.

MichaIng replied:
Ah, I didn't think about the values shown in the menu afterwards. Currently this is required to not have HTTP authentication used.

But then we need to a separate variable for this:
```
local http_auth=1
...
if [[ $PROVIDER == 'DuckDNS' ]]
then
	unset -v http_auth
...
...${http_auth:+ -u '$USERNAME:$PASSWORD'}...
```
Will implement this now.

MichaIng replied:
Done: https://github.com/MichaIng/DietPi/commit/5ce84c7ed90ffd6c882e18ee2f2f3c2854647838
